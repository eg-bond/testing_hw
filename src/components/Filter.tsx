import { useEffect } from 'react';
import { useSelector } from 'react-redux';
import { activeTasksSelector, doneTasksSelector } from 'src/store/taskSlice';

export enum FilterType {
  All = 'all',
  Active = 'active',
  Done = 'done',
}

interface IFilter {
  currentFilter: FilterType;
  setItemsToList: React.Dispatch<React.SetStateAction<Task[]>>;
  setCurrentFilter: React.Dispatch<React.SetStateAction<FilterType>>;
  allTasks: Task[];
}

export const Filter = ({
  currentFilter,
  setItemsToList,
  setCurrentFilter,
  allTasks,
}: IFilter) => {
  const activeTasks = useSelector(activeTasksSelector);
  const doneTasks = useSelector(doneTasksSelector);
  // update itemsToList if state changes
  useEffect(() => {
    if (currentFilter === FilterType.All) setItemsToList(allTasks);
    if (currentFilter === FilterType.Active) setItemsToList(activeTasks);
    if (currentFilter === FilterType.Done) setItemsToList(doneTasks);
  }, [allTasks, activeTasks, doneTasks]);

  const handleFilter = (filterType: FilterType) => {
    if (filterType === currentFilter) {
      setItemsToList(allTasks);
      setCurrentFilter(FilterType.All);
      return;
    }
    if (filterType === FilterType.Active) {
      setItemsToList(activeTasks);
    }
    if (filterType === FilterType.Done) {
      setItemsToList(doneTasks);
    }
    setCurrentFilter(filterType);
  };

  return (
    <div>
      <button
        style={{
          color: 'white',
          backgroundColor:
            currentFilter === FilterType.Active ? 'red' : '#293547',
          borderRadius: '7px',
          padding: '0.5rem',
        }}
        onClick={() => handleFilter(FilterType.Active)}>
        Активные
      </button>
      <button
        style={{
          color: 'white',
          backgroundColor:
            currentFilter === FilterType.Done ? 'red' : '#293547',
          borderRadius: '7px',
          padding: '0.5rem',
        }}
        onClick={() => handleFilter(FilterType.Done)}>
        Выполненные
      </button>
    </div>
  );
};
