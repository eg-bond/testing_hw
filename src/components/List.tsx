import { Item } from './Item';

type Props = {
  items: Task[];
  onDelete: (id: Task['id']) => void;
  onToggle: (id: Task['id']) => void;
};

const MAX_ITEMS_TO_SHOW = 10;

export const List = ({ items, onDelete, onToggle }: Props) => {
  const activeTasks = items.filter(item => !item.done);
  if (activeTasks.length > MAX_ITEMS_TO_SHOW) {
    items = items.slice(0, 10);
  }

  return (
    <>
      <ul className='task-list tasks'>
        {items.map(item => (
          <Item
            {...item}
            key={item.id}
            onDelete={onDelete}
            onToggle={onToggle}
          />
        ))}
      </ul>
      {activeTasks.length > MAX_ITEMS_TO_SHOW && (
        <div className='empty-wrapper'>
          В списке есть еще {activeTasks.length - 10} активных задачи, но вы не
          можете их видеть, так как их слишком много. <br />
          Показаны первые 10 задач.
        </div>
      )}
    </>
  );
};
