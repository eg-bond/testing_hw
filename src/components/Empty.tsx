import { FilterType } from './Filter';

interface IEmpty {
  allTasks: Task[];
}

export const Empty = ({ allTasks }: IEmpty) => {
  if (allTasks.length > 0) {
    return null;
  }
  return (
    <div className='empty-wrapper'>
      Вы пока не создали ни одной задачи. <br />
      Давайте что-нибудь запланируем?
    </div>
  );
};
