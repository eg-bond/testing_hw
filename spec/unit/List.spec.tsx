import { render, screen } from '@testing-library/react';
import { List } from 'src/components/List';

const generateUndoneTasks = (count: number): Task[] => {
  return Array.from({ length: count }, (_, index) => ({
    id: (index + 1).toString(),
    header: '...',
    done: false,
  }));
};

it('отображение списка задач', () => {
  const onDelete = jest.fn();
  const onToggle = jest.fn();

  const items = generateUndoneTasks(3);

  const { rerender, asFragment } = render(
    <List items={items} onDelete={onDelete} onToggle={onToggle} />
  );
  const firstRender = asFragment();

  items.pop();

  rerender(<List items={items} onDelete={onDelete} onToggle={onToggle} />);
  const secondRender = asFragment();

  expect(firstRender).toMatchDiffSnapshot(secondRender);
});

it('Список содержит не больше 10 невыполненных задач', () => {
  const onDelete = jest.fn();
  const onToggle = jest.fn();

  const items = generateUndoneTasks(13);

  render(<List items={items} onDelete={onDelete} onToggle={onToggle} />);

  const shownItems = screen.getAllByRole('listitem');

  expect(shownItems.length).toBeLessThanOrEqual(10);
});

it('Список выводит уведомление о том, что невыполненных задач больше чем 10', () => {
  const onDelete = jest.fn();
  const onToggle = jest.fn();

  const items = generateUndoneTasks(10);

  const { rerender } = render(
    <List items={items} onDelete={onDelete} onToggle={onToggle} />
  );

  const absentTooMuchTasksWarning = screen.queryByText(/В списке есть еще/i);
  expect(absentTooMuchTasksWarning).not.toBeInTheDocument();

  items.push({
    id: '11',
    header: '...',
    done: false,
  });

  rerender(<List items={items} onDelete={onDelete} onToggle={onToggle} />);

  const tooMuchTasksWarning = screen.getByText(/В списке есть еще/i);
  expect(tooMuchTasksWarning).toBeInTheDocument();
});
