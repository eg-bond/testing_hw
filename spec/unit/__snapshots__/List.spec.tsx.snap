// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`отображение списка задач 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -46,30 +46,7 @@
            alt="Удалить"
            src="[object Object]"
          />
        </button>
      </li>
-     <li
-       class="item-wrapper"
-     >
-       <input
-         id="3"
-         type="checkbox"
-       />
-       <label
-         for="3"
-       >
-         ...
-       </label>
-       <button
-         class="button button-with-icon"
-         data-alt="Удалить"
-         disabled=""
-       >
-         <img
-           alt="Удалить"
-           src="[object Object]"
-         />
-       </button>
-     </li>
    </ul>
  </DocumentFragment>"
`;
